File: apps/desktop/src/ipc/messages.ts
import type {
  CoreFlowReason,
  LogCode,
  LogLineBody,
  NoticeCode,
  SpeedTestOutcome,
  ValidationCode,
  ValidationIssue,
  ValidationScope,
} from "@/ipc/bindings";
import type { TranslationFunction, TranslationKey } from "@voya/i18n";

/**
 * The one place a backend message code becomes text.
 *
 * The backend used to hand the frontend finished English sentences — notice
 * titles, validator messages, speedtest statuses, log lines — which the eight
 * shipped locales could not touch and `pnpm check:i18n` could not see, because
 * it only scans frontend source. Every one of those is a code now, and these
 * registries map it to the locale key that spells it out.
 *
 * Each map is `Record<Code, TranslationKey>`, so both halves are checked:
 * `tsc` fails when a Rust code has no entry (the generated union gained a
 * member) and when an entry names a key `en.json` does not define.
 *
 * The interpolation payload is the code object itself. `ValidationCode` and
 * friends are internally tagged, so `{ code: "unsupportedNetwork", network:
 * "kcp" }` already *is* the bag i18next reads `{{network}}` out of; there is no
 * second parameter list to keep in step.
 */
export const NOTICE_KEYS: Record<NoticeCode["code"], TranslationKey> = {
  configurationRefreshFailed: "notices.configurationRefreshFailed",
  connectionModeRefreshFailed: "notices.connectionModeRefreshFailed",
  connectionModeSavedRestartFailed: "notices.connectionModeSavedRestartFailed",
  coreStartedSystemProxyFailed: "notices.coreStartedSystemProxyFailed",
  coreStopped: "notices.coreStopped",
  dnsRefreshFailed: "notices.dnsRefreshFailed",
  dnsSavedRestartFailed: "notices.dnsSavedRestartFailed",
  nativeTunStopped: "notices.nativeTunStopped",
  profileRefreshFailed: "notices.profileRefreshFailed",
  proxyModeSavedRuntimeUpdateFailed: "notices.proxyModeSavedRuntimeUpdateFailed",
  proxyViewRefreshFailed: "notices.proxyViewRefreshFailed",
  routingDeletedRestartFailed: "notices.routingDeletedRestartFailed",
  routingRefreshFailed: "notices.routingRefreshFailed",
  routingRuleMovedRestartFailed: "notices.routingRuleMovedRestartFailed",
  routingRuleSavedRestartFailed: "notices.routingRuleSavedRestartFailed",
  routingRulesDeletedRestartFailed: "notices.routingRulesDeletedRestartFailed",
  routingSavedRestartFailed: "notices.routingSavedRestartFailed",
  routingSelectedRestartFailed: "notices.routingSelectedRestartFailed",
  settingsRefreshFailed: "notices.settingsRefreshFailed",
  settingsSavedRuntimeUpdateFailed: "notices.settingsSavedRuntimeUpdateFailed",
  settingsSavedSystemProxyUpdateFailed: "notices.settingsSavedSystemProxyUpdateFailed",
  subscriptionAutoUpdateFailed: "notices.subscriptionAutoUpdateFailed",
  subscriptionRefreshFailed: "notices.subscriptionRefreshFailed",
  systemProxyRestoreFailed: "notices.systemProxyRestoreFailed",
  systemProxyStatusRefreshFailed: "notices.systemProxyStatusRefreshFailed",
  templateImportedRestartFailed: "notices.templateImportedRestartFailed",
  trayRefreshFailed: "notices.trayRefreshFailed",
  tunSavedRestartFailed: "notices.tunSavedRestartFailed",
  tunStatusRefreshFailed: "notices.tunStatusRefreshFailed",
};

export const LOG_KEYS: Record<LogCode["code"], TranslationKey> = {
  connected: "logCodes.connected",
  connecting: "logCodes.connecting",
  coreExitGaveUp: "logCodes.coreExitGaveUp",
  coreExitRestarted: "logCodes.coreExitRestarted",
  coreExitRetryScheduled: "logCodes.coreExitRetryScheduled",
  coreOperationFailed: "logCodes.coreOperationFailed",
  disconnected: "logCodes.disconnected",
  disconnecting: "logCodes.disconnecting",
  nativeTunExited: "logCodes.nativeTunExited",
  postCommitFailed: "logCodes.postCommitFailed",
  previousCoreStillRunning: "logCodes.previousCoreStillRunning",
  restarted: "logCodes.restarted",
  restartedAfterChange: "logCodes.restartedAfterChange",
  restarting: "logCodes.restarting",
  restartingAfterChange: "logCodes.restartingAfterChange",
  runtimeStatusRefreshFailed: "logCodes.runtimeStatusRefreshFailed",
  speedtestCancellationRequested: "logCodes.speedtestCancellationRequested",
  subscriptionAutoUpdateFailed: "logCodes.subscriptionAutoUpdateFailed",
  subscriptionAutoUpdateFinished: "logCodes.subscriptionAutoUpdateFinished",
};

/** The operation a log sentence is about, interpolated into it as `reason`. */
export const CORE_FLOW_REASON_KEYS: Record<CoreFlowReason, TranslationKey> = {
  configTemplateImported: "coreFlowReason.configTemplateImported",
  connect: "coreFlowReason.connect",
  connectionModeChanged: "coreFlowReason.connectionModeChanged",
  disconnect: "coreFlowReason.disconnect",
  dnsChanged: "coreFlowReason.dnsChanged",
  restart: "coreFlowReason.restart",
  routingChanged: "coreFlowReason.routingChanged",
  settingsSaved: "coreFlowReason.settingsSaved",
  tunChanged: "coreFlowReason.tunChanged",
};

export const VALIDATION_KEYS: Record<ValidationCode["code"], TranslationKey> = {
  dnsAddressEmpty: "validation.dnsAddressEmpty",
  dnsAddressPort: "validation.dnsAddressPort",
  dnsExpectedIps: "validation.expectedIps",
  dnsHostsLine: "validation.dnsHostsLine",
  groupChildNotFound: "validation.groupChildNotFound",
  groupCycle: "validation.groupCycle",
  groupCyclePath: "validation.groupCyclePath",
  groupDuplicateChildIgnored: "validation.groupDuplicateChildIgnored",
  groupWithoutValidChild: "validation.groupWithoutValidChild",
  hysteriaHopIntervalTooShort: "validation.hysteriaHopIntervalTooShort",
  invalidAddress: "validation.invalidAddress",
  invalidFinalMask: "validation.invalidFinalMask",
  invalidFlow: "validation.invalidFlow",
  invalidPassword: "validation.invalidPassword",
  invalidPort: "validation.invalidPort",
  invalidRealityPublicKey: "validation.invalidRealityPublicKey",
  invalidShadowsocksMethod: "validation.invalidShadowsocksMethod",
  invalidSubscriptionFilter: "validation.invalidSubscriptionFilter",
  invalidUdpTestTarget: "validation.invalidUdpTestTarget",
  negativeHysteriaBandwidth: "validation.negativeHysteriaBandwidth",
  notAGroupProfile: "validation.notAGroupProfile",
  policyGroupWithoutValidChildren: "validation.policyGroupWithoutValidChildren",
  proxyChainSingleHop: "validation.proxyChainSingleHop",
  proxyChainWithoutValidChildren: "validation.proxyChainWithoutValidChildren",
  routingRuleOutboundNotFound: "validation.routingRuleOutboundNotFound",
  routingRuleWithoutOutbound: "validation.routingRuleWithoutOutbound",
  sourceUrlHasCredentials: "validation.urlCredentials",
  sourceUrlNotHttp: "validation.urlInvalid",
  sourceUrlNotHttps: "validation.urlHttps",
  textControlCharacters: "validation.textControlCharacters",
  textRequired: "validation.textRequired",
  textTooLong: "validation.textTooLong",
  tooManyItems: "validation.tooManyItems",
  tunMtuOutOfRange: "validation.tunMtuOutOfRange",
  unsupportedNetwork: "validation.unsupportedNetwork",
  unsupportedProtocol: "validation.unsupportedProtocol",
  unsupportedProtocolNetwork: "validation.unsupportedProtocolNetwork",
  unsupportedSettingsSchema: "validation.unsupportedSettingsSchema",
  unsupportedShadowsocksNetwork: "validation.unsupportedShadowsocksNetwork",
  // Rendered from its own English `message`, never through a key.
  untranslated: "validation.invalid",
};

export const VALIDATION_SCOPE_KEYS: Record<ValidationScope["kind"], TranslationKey> = {
  groupChild: "validation.scope.groupChild",
  routingRuleOutbound: "validation.scope.routingRuleOutbound",
};

export const SPEEDTEST_OUTCOME_KEYS: Record<SpeedTestOutcome, TranslationKey> = {
  cancelled: "speedtest.outcome.cancelled",
  completed: "speedtest.outcome.completed",
  coreUnavailable: "speedtest.outcome.coreUnavailable",
  failed: "speedtest.outcome.failed",
  invalidProfile: "speedtest.outcome.invalidProfile",
  noAvailablePort: "speedtest.outcome.noAvailablePort",
  proxyConnectFailed: "speedtest.outcome.proxyConnectFailed",
  proxyConnectionClosed: "speedtest.outcome.proxyConnectionClosed",
  proxyConnectionRefused: "speedtest.outcome.proxyConnectionRefused",
  skipped: "speedtest.outcome.skipped",
  testing: "speedtest.outcome.testing",
  timedOut: "speedtest.outcome.timedOut",
  udpTestFailed: "speedtest.outcome.udpTestFailed",
  unknown: "speedtest.outcome.unknown",
  waiting: "speedtest.outcome.waiting",
};

export function noticeText(t: TranslationFunction, code: NoticeCode) {
  return t(NOTICE_KEYS[code.code], code);
}

/**
 * One log line as the panel shows it.
 *
 * Core output and `tracing` diagnostics are passed through; only the app's own
 * sentences are translated, with their untranslated `detail` after them.
 */
export function logLineText(t: TranslationFunction, body: LogLineBody) {
  if (body.source !== "app") {
    return body.line;
  }

  const reason = "reason" in body.code ? t(CORE_FLOW_REASON_KEYS[body.code.reason]) : undefined;
  const text = t(LOG_KEYS[body.code.code], { ...body.code, reason });

  return body.detail ? `${text}: ${body.detail}` : text;
}

/**
 * One validation finding, with the breadcrumb the validator walked to reach it.
 *
 * `untranslated` is the escape hatch for managers that have no code yet, and
 * shows their English diagnostic rather than a locale key nobody wrote.
 */
export function validationText(t: TranslationFunction, issue: ValidationIssue) {
  const message =
    issue.code.code === "untranslated"
      ? issue.code.message
      : t(VALIDATION_KEYS[issue.code.code], {
          ...issue.code,
          ...("path" in issue.code ? { path: issue.code.path.join(" → ") } : {}),
        });
  if (issue.scope.length === 0) {
    return message;
  }

  const breadcrumb = issue.scope
    .map((scope) => t(VALIDATION_SCOPE_KEYS[scope.kind], scope))
    .join(" / ");

  return `${breadcrumb}: ${message}`;
}

export function speedtestOutcomeText(t: TranslationFunction, outcome: SpeedTestOutcome) {
  return t(SPEEDTEST_OUTCOME_KEYS[outcome]);
}

File: crates/voya-contracts/src/messages.rs
//! Machine codes for the text a user reads.
//!
//! The app gates locale alignment in CI, but that gate
//! only sees `apps/desktop/src` and `packages/ui/src`. Any sentence built in
//! Rust therefore reached the screen in English whatever the interface
//! language was — notice titles, validator messages, tray labels, speedtest
//! status text — and three of those were persisted into SQLite, so switching
//! the language later could not fix them either.
//!
//! Every user-visible string the backend produces now crosses the boundary as
//! a **code plus its parameters**, and the frontend resolves it against the
//! locale files (`apps/desktop/src/ipc/messages.ts` holds the code → key
//! registries). The enums are internally tagged, so one payload is both the
//! discriminator and the interpolation bag: the frontend calls
//! `t(KEY_BY_CODE[value.code], value)` and i18next reads `{{network}}` out of
//! the very same object.
//!
//! What deliberately stays English:
//!
//! * `detail` fields. They carry an error's `Display` text for the logs and
//!   the "why" line under a notice — the same contract `AppError::message`
//!   already has. They are never parsed and never persisted.
//! * `LogLineBody::Core`. Those bytes are the core process's own output.

use serde::{Deserialize, Serialize};
use specta::Type;

/// What the backend is telling the user about, as a code the frontend
/// translates.
///
/// One variant per distinct message rather than a message assembled from
/// parts: a notice is a whole sentence in every locale, and languages do not
/// agree on how to build one out of a subject and a verb.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(
    tag = "code",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum NoticeCode {
    // ---- a committed change could not be broadcast to the query caches ----
    ProfileRefreshFailed,
    SubscriptionRefreshFailed,
    RoutingRefreshFailed,
    DnsRefreshFailed,
    ProxyViewRefreshFailed,
    ConnectionModeRefreshFailed,
    SettingsRefreshFailed,
    // ---- the change was saved but the core could not be restarted on it ----
    RoutingSavedRestartFailed,
    RoutingDeletedRestartFailed,
    RoutingSelectedRestartFailed,
    RoutingRuleSavedRestartFailed,
    RoutingRulesDeletedRestartFailed,
    RoutingRuleMovedRestartFailed,
    DnsSavedRestartFailed,
    TunSavedRestartFailed,
    ConnectionModeSavedRestartFailed,
    SettingsSavedRuntimeUpdateFailed,
    ProxyModeSavedRuntimeUpdateFailed,
    // ---- the change was saved but a follow-up side effect failed ----
    SettingsSavedSystemProxyUpdateFailed,
    SystemProxyStatusRefreshFailed,
    TunStatusRefreshFailed,
    TrayRefreshFailed,
    // ---- the core or the tunnel stopped on its own ----
    CoreStopped,
    NativeTunStopped,
    CoreStartedSystemProxyFailed,
    SystemProxyRestoreFailed,
    // ---- background work ----
    SubscriptionAutoUpdateFailed { remarks: String },
}

/// Which core operation a log line is about.
///
/// The flow used to interpolate an English fragment ("connect", "Routing
/// changed") into its log sentences; the fragment is a code now so the whole
/// sentence can be assembled in the reader's language.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(rename_all = "camelCase")]
pub enum CoreFlowReason {
    Connect,
    Restart,
    Disconnect,
    RoutingChanged,
    DnsChanged,
    TunChanged,
    ConnectionModeChanged,
    SettingsSaved,
}

/// A log line the **app** wrote, named by code.
///
/// Only lines the app authors are in here. Everything the core process prints
/// travels as [`LogLineBody::Core`] and is shown byte for byte.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(
    tag = "code",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum LogCode {
    Connecting,
    Connected,
    Restarting,
    Restarted,
    Disconnecting,
    Disconnected,
    RestartingAfterChange {
        reason: CoreFlowReason,
    },
    RestartedAfterChange {
        reason: CoreFlowReason,
    },
    /// The core exited and was restarted straight away.
    CoreExitRestarted {
        attempt: u32,
    },
    /// The core exited and a restart is scheduled.
    CoreExitRetryScheduled {
        attempt: u32,
        delay_ms: u32,
    },
    /// The core exited too often and the app stopped restarting it.
    CoreExitGaveUp,
    NativeTunExited,
    /// An operation failed before the supervisor was touched, so the core that
    /// was already running is still serving traffic.
    PreviousCoreStillRunning {
        reason: CoreFlowReason,
    },
    /// The supervisor could not say what state it ended up in.
    RuntimeStatusRefreshFailed {
        reason: CoreFlowReason,
    },
    /// A core operation failed; `detail` carries the error.
    CoreOperationFailed {
        reason: CoreFlowReason,
    },
    /// A change was committed and its follow-up work failed.
    PostCommitFailed,
    SpeedtestCancellationRequested,
    SubscriptionAutoUpdateFailed {
        remarks: String,
    },
    SubscriptionAutoUpdateFinished {
        remarks: String,
        imported: u32,
    },
}

/// One line in the Logs panel.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(
    tag = "source",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum LogLineBody {
    /// Output of the core process, passed through unchanged. It is the core's
    /// own text in the core's own wording, so translating it is not ours to do.
    Core { line: String },
    /// A `tracing` event from the app's own instrumentation, forwarded to the
    /// panel so a failure is visible without opening the file log. Structured
    /// developer diagnostics with a module target — deliberately untranslated,
    /// for the same reason a stack trace is.
    Diagnostic { line: String },
    /// A line the app wrote *for the user*. `detail` is an untranslated
    /// technical diagnostic (an error's `Display`), shown after the translated
    /// sentence.
    App {
        code: LogCode,
        detail: Option<String>,
    },
}

/// Why a submitted or generated value was rejected.
///
/// Produced by `voya_core`'s node/group validators, by the DNS pane's checks
/// and by settings validation. [`ValidationCode::Untranslated`] is the
/// deliberate escape hatch for the managers that have no code yet: it is
/// greppable, it shows the English diagnostic verbatim, and adding a code for
/// one is a purely additive change.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(
    tag = "code",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum ValidationCode {
    SubscriptionReadOnly {
        subscription_id: String,
    },
    // ---- node fields (voya_core::context::validation) ----
    InvalidAddress,
    InvalidPort,
    InvalidPassword,
    InvalidFlow,
    InvalidShadowsocksMethod,
    InvalidRealityPublicKey,
    UnsupportedProtocol {
        protocol: String,
    },
    UnsupportedProtocolNetwork {
        protocol: String,
        network: String,
    },
    UnsupportedShadowsocksNetwork {
        network: String,
    },
    // ---- routing rules ----
    RoutingRuleWithoutOutbound {
        rule: String,
    },
    RoutingRuleOutboundNotFound {
        rule: String,
        outbound: String,
    },
    // ---- DNS ----
    DnsAddressEmpty,
    DnsAddressPort {
        port: String,
    },
    DnsHostsLine {
        line: u32,
    },
    DnsExpectedIps,
    // ---- application settings ----
    TextRequired,
    TextTooLong,
    TextControlCharacters,
    TooManyItems,
    UnsupportedSettingsSchema {
        found: u32,
        expected: u32,
    },
    TunMtuOutOfRange {
        min: u32,
        max: u32,
    },
    NegativeHysteriaBandwidth,
    HysteriaHopIntervalTooShort {
        minimum_seconds: u32,
    },
    /// A rejection this contract has no code for. The English `message` is the
    /// failing manager's own diagnostic and is rendered verbatim.
    Untranslated {
        message: String,
    },
}

/// Where a validation message was produced, when the validator had to walk
/// into a group's children or a routing rule's outbound to find it.
///
/// This replaces the `"group child A / B: "` prefixes the validators used to
/// glue onto the front of a message, which made the message itself
/// untranslatable. The frontend renders the breadcrumb from these entries and
/// the translated message after it.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(
    tag = "kind",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum ValidationScope {
    RoutingRuleOutbound { rule: String, outbound: String },
}

/// One rejected value.
///
/// `field` is a stable identifier, not a display label: the DNS pane keys its
/// inputs by `direct`/`remote`/`bootstrap`/`hosts`, the settings surface by its
/// contract path (`network.tun.mtu`), and the group builder by `children`.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ValidationIssue {
    pub field: String,
    pub code: ValidationCode,
    /// Empty unless the validator descended into a group or a rule outbound.
    pub scope: Vec<ValidationScope>,
}

impl ValidationIssue {
    #[must_use]
    pub fn new(field: impl Into<String>, code: ValidationCode) -> Self {
        Self {
            field: field.into(),
            code,
            scope: Vec::new(),
        }
    }

    /// An issue with no code of its own, carrying the manager's English text.
    #[must_use]
    pub fn untranslated(field: impl Into<String>, message: impl Into<String>) -> Self {
        Self::new(
            field,
            ValidationCode::Untranslated {
                message: message.into(),
            },
        )
    }
}

/// Why one line of imported text did not simply become a node.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(
    tag = "code",
    rename_all = "camelCase",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum ImportLineCode {
    /// The line was a subscription URL and was added as a source instead.
    SubscriptionSourceAdded,
    /// The share link names a transport sing-box cannot carry.
    UnsupportedTransport { transport: String },
    /// Any other share-link parse failure, with its untranslated diagnostic.
    ParseFailed { detail: String },
}

/// One reported line of an import, numbered from 1.
#[derive(Debug, Clone, PartialEq, Eq, Deserialize, Serialize, Type)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ImportLineIssue {
    pub line: u32,
    pub code: ImportLineCode,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn codes_serialize_as_camel_case_tagged_objects() {
        let value = serde_json::to_value(NoticeCode::SubscriptionAutoUpdateFailed {
            remarks: "Feed".to_string(),
        })
        .expect("serialize notice code");

        assert_eq!(value["code"], "subscriptionAutoUpdateFailed");
        assert_eq!(value["remarks"], "Feed");
    }

    #[test]
    fn log_line_body_separates_core_output_from_app_authored_lines() {
        let core = serde_json::to_value(LogLineBody::Core {
            line: "inbound/mixed".to_string(),
        })
        .expect("serialize core line");
        let app = serde_json::to_value(LogLineBody::App {
            code: LogCode::RestartingAfterChange {
                reason: CoreFlowReason::DnsChanged,
            },
            detail: None,
        })
        .expect("serialize app line");

        assert_eq!(core["source"], "core");
        assert_eq!(core["line"], "inbound/mixed");
        assert_eq!(app["source"], "app");
        assert_eq!(app["code"]["code"], "restartingAfterChange");
        assert_eq!(app["code"]["reason"], "dnsChanged");
    }

    #[test]
    fn validation_issue_carries_its_breadcrumb() {
        let issue = ValidationIssue {
            field: "activeProfile".to_string(),
            code: ValidationCode::InvalidPort,
            scope: vec![ValidationScope::RoutingRuleOutbound {
                rule: "Group".to_string(),
                outbound: "Leaf".to_string(),
            }],
        };
        let value = serde_json::to_value(&issue).expect("serialize issue");

        assert_eq!(value["field"], "activeProfile");
        assert_eq!(value["code"]["code"], "invalidPort");
        assert_eq!(value["scope"][0]["kind"], "routingRuleOutbound");
        assert_eq!(value["scope"][0]["rule"], "Group");
    }

    #[test]
    fn untranslated_keeps_the_managers_own_diagnostic() {
        let issue = ValidationIssue::untranslated("pem", "certificate is not valid PEM");

        assert_eq!(
            issue.code,
            ValidationCode::Untranslated {
                message: "certificate is not valid PEM".to_string()
            }
        );
    }
}
